### DISCUSSION TITLE
Incorrect result of multiple compound assignment operator

### DISCUSSION THREAD

**Discussion by jip** [dotnet/csharplang#7741] [General] [answered]
The following code should swap values in a and b:

```c#
int a = 3, b = 5;
a ^= b ^= a ^= b;
Console.WriteLine($"a = {a}, b = {b}");
```
but it gives wrong result:
`a = 0, b = 3`

But the code with separated assignments works:
```c#
int a = 3, b = 5;
a ^= b;
b ^= a;
a ^= b;
Console.WriteLine($"a = {a}, b = {b}");
```
producing:
`a = 5, b = 3`

Environment:
```console
C:\Program Files\dotnet>dotnet --version
8.0.100

C:\Program Files\dotnet>dotnet --list-sdks
8.0.100 [C:\Program Files\dotnet\sdk]

C:\Program Files\dotnet>dotnet --list-runtimes
Microsoft.AspNetCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
Microsoft.NETCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
Microsoft.WindowsDesktop.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

**Comment by AndersKehlet** [comment]
Well, that was a fun little rabbit hole.

Of the langs I tested, C++ and PHP do what you want
Go, Swift and Python don't seem to allow assignments as expressions at all.
JS and Java behave like C#.

If you look at the IL it's easy to see _why_ it behaves like that in C#, and I'd very strongly assume it's a deliberate choice.

**Reply by AndersKehlet** [reply]
Current IL:

```
ldc.i4.3
stloc.0      // a
ldc.i4.5
stloc.1      // b

ldloc.0      // a
ldloc.1      // b
ldloc.0      // a
ldloc.1      // b
<<op>>
dup
stloc.0      // a
<<op>>
dup
stloc.1      // b
<<op>>
stloc.0
```

I assume the correct IL would be something like this:

```
ldc.i4.3
stloc.0      // a
ldc.i4.5
stloc.1      // b

ldloc.1      // b
ldloc.0      // a
ldloc.1      // b
<<op>>
dup
stloc.0      // a
<<op>>
dup
stloc.1      // b
ldloc.0      // a
<<op>>
stloc.0
```

**Comment by HaloFour** [comment]
This seems like a pretty esoteric case, and the language has probably done this since C# 1.0.  I guess the question is what the language specification states that the language should do and whether it's worth amending.  If the language spec implies that this should work as you described, then I would post to the [Roslyn](https://github.com/dotnet/roslyn) repository as a compiler bug.

**Reply by huoyaoyuan** [reply]
I'd rather like to see the language specifies this as an error. This is unlikely to happen though, because of breaking and detailed ruling.

**Comment by jip** [comment]
1. The result of assignment expression is an assigned value ([§12.2.2](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#1222-values-of-expressions)).
2. Assignment operators are right-associative ([§12.4.2](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#1242-operator-precedence-and-associativity), [§12.21.1](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#12211-general)).
3. The compound assignment operator `x «op»= y` is permitted if both of `x «op» y` and `x = y` are permitted ([12.21.4](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#12214-compound-assignment)).

So, the expression `a ^= b ^= a ^= b;` is valid in C#.

I don't see statements kinda "correct work of esoteric cases is not guaranteed" or "when in doubt, consult with C++/PHP/whatever language agreements" or "correct work of associative operators in chain is guaranteed only for chains of length not more than 2" (because `b ^= a ^= b;` works correctly) in C# Language Reference.

I'll report my previous post to Roslyn repo, thanks all.

**Reply by AndersKehlet** [reply]
I assure you there's no need for sarcasm. :)

**Comment by jip** [comment]
https://github.com/dotnet/roslyn/issues/71139

**Comment by jip** [comment, marked as answer]
The explanation why the result observed is as intended see in Roslyn thread.
